File: elastic-tube-1d/solid-cpp/src/SolidComputeSolution.cpp
#include <cmath>
#include "SolidSolver.h"

void SolidComputeSolution(int chunkLength, double const *const pressure, double *crossSectionLength)
{
  /*
   * Update displacement of membrane based on pressure data from the fluid solver
   */
  const double PI        = 3.14159265359;
  const double E         = 10000;
  const double r0        = 1.0 / std::sqrt(PI);
  const double c_mk      = std::sqrt(E / 2 / r0);
  const double c_mk2     = c_mk * c_mk;
  const double pressure0 = 0.0;

  for (int i = 0; i < chunkLength; i++) {
    crossSectionLength[i] = std::pow((pressure0 - 2.0 * c_mk2), 2.0) / std::pow((pressure[i] - 2.0 * c_mk2), 2.0);
  }
}

File: elastic-tube-1d/solid-fortran-module/src/SolidComputeSolution.f90
module SolidComputeSolution
  implicit none
  integer, parameter :: dp = kind(1.0d0)

contains

  subroutine solid_compute_solution(chunkLength, pressure, crossSectionLength)
    integer, intent(in) :: chunkLength
    real(dp), intent(in) :: pressure(1:chunkLength)
    real(dp), intent(inout) :: crossSectionLength(1:chunkLength)

    real(dp) :: pi, e, r0, c_mk, c_mk2
    real(dp) :: pressure0
    integer :: i

    ! constants
    pi = 3.141592653589793_dp
    e = 10000.0_dp
    r0 = 1.0_dp / sqrt(pi)
    c_mk = sqrt(e / (2.0_dp * r0))
    c_mk2 = c_mk * c_mk
    pressure0 = 0.0_dp

    ! Update crossSectionLength based on pressure
    do i = 1, chunkLength
      crossSectionLength(i) = ((pressure0 - 2.0_dp * c_mk2)**2) / &
                              ((pressure(i) - 2.0_dp * c_mk2)**2)
    end do

  end subroutine solid_compute_solution

end module SolidComputeSolution
